test/e2e/e2e_setup_test.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/types"
+	. "sigs.k8s.io/cluster-api/test/framework/ginkgoextensions"
+	"sigs.k8s.io/yaml"
+)
+
+type setupOptions struct {
+	additionalIPVariableNames []string
+}
+
+// SetupOption is a configuration option supplied to Setup.
+type SetupOption func(*setupOptions)
+
+// WithIP instructs Setup to allocate another IP and store it into the provided variableName
+// NOTE: Setup always allocate an IP for CONTROL_PLANE_ENDPOINT_IP.
+func WithIP(variableName string) SetupOption {
+	return func(o *setupOptions) {
+		o.additionalIPVariableNames = append(o.additionalIPVariableNames, variableName)
+	}
+}
+
+// Setup for the specific test.
+func Setup(name string, f func(string), opts ...SetupOption) {
+	options := &setupOptions{}
+	for _, o := range opts {
+		o(options)
+	}
+
+	var (
+		testSpecificClusterctlConfigPath string
+		testSpecificIPAddressClaims      []types.NamespacedName
+		testSpecificVariables            map[string]string
+	)
+	BeforeEach(func() {
+		It("Should prepare test specif setup", func() {
```

Comment:
Should this be a By?
```suggestion
		By("Should prepare test specif setup", func() {
```

```
  Ginkgo detected an issue with your spec structure
  It("Should prepare test specif setup", func() {
  /home/prow/go/src/sigs.k8s.io/cluster-api-provider-vsphere/test/e2e/e2e_setup_test.go:60
    It looks like you are trying to add a [It] node
    to the Ginkgo spec tree in a leaf node after the specs started running.
    To enable randomization and parallelization Ginkgo requires the spec tree
    to be fully constructed up front.  In practice, this means that you can
    only create nodes like [It] at the top-level or within the
    body of a Describe, Context, or When.
    Learn more at:
    http://onsi.github.io/ginkgo/#mental-model-how-ginkgo-traverses-the-spec-hierarchy
```

I read this as `It` cannot be in a `BeforeEach` or `AfterEach`.